File: module-18-network-forensics/incident_timeline.js
/**
 * Module 18 — JS/Node port of incident_timeline.py. Same recon/exploit/
 * exfiltration detection and chronological reconstruction.
 */
const { readPcap, parseFrame } = require("./pcap_reader.js");

const KNOWN_BAD_PAYLOADS = new Map([
  ["smiley:)", "vsftpd 2.3.4 backdoor trigger string (CVE-2011-2523, same CVE as Module 16's database)"],
]);

function annotatePacket(parsed) {
  const payloadStr = parsed.payload.toString();
  for (const [pattern, description] of KNOWN_BAD_PAYLOADS) {
    if (payloadStr.includes(pattern)) {
      return `EXPLOIT ATTEMPT -- payload matches known-bad pattern: ${description}`;
    }
  }
  return null;
}

function detectRecon(packetsParsed, windowSeconds = 1.0, minPorts = 4) {
  const bySrcDst = new Map();
  for (const [ts, p] of packetsParsed) {
    if (p.tcpFlags === 0x02) {
      const key = `${p.srcIp}|${p.dstIp}`;
      if (!bySrcDst.has(key)) bySrcDst.set(key, []);
      bySrcDst.get(key).push([ts, p.dstPort]);
    }
  }

  const findings = [];
  for (const [key, events] of bySrcDst) {
    const [src, dst] = key.split("|");
    events.sort((a, b) => a[0] - b[0]);
    const ports = new Set(events.filter(([ts]) => ts - events[0][0] <= windowSeconds).map(([, port]) => port));
    if (ports.size >= minPorts) {
      findings.push({ type: "reconnaissance", src, dst, portsTouched: [...ports].sort((a, b) => a - b), startTs: events[0][0] });
    }
  }
  return findings;
}

function detectExfiltration(packetsParsed, minPackets = 10) {
  const bySrcDstPort = new Map();
  for (const [ts, p] of packetsParsed) {
    if (p.payload.length > 0) {
      const key = `${p.srcIp}|${p.dstIp}|${p.dstPort}`;
      if (!bySrcDstPort.has(key)) bySrcDstPort.set(key, []);
      bySrcDstPort.get(key).push([ts, p.payload.length]);
    }
  }

  const findings = [];
  for (const [key, events] of bySrcDstPort) {
    if (events.length >= minPackets) {
      const [src, dst, port] = key.split("|");
      const totalBytes = events.reduce((sum, [, size]) => sum + size, 0);
      findings.push({ type: "exfiltration", src, dst, port: Number(port), packetCount: events.length, totalBytes, startTs: events[0][0] });
    }
  }
  return findings;
}

function buildTimeline(path) {
  const { packets } = readPcap(path);
  const packetsParsed = packets.map((pkt) => [pkt.ts, parseFrame(pkt.frame)]);

  const events = [];

  for (const f of detectRecon(packetsParsed)) {
    events.push([f.startTs, `RECONNAISSANCE: ${f.src} scanned ${f.portsTouched.length} ` +
      `ports on ${f.dst} ([${f.portsTouched.join(", ")}])`]);
  }

  for (const [ts, p] of packetsParsed) {
    const note = annotatePacket(p);
    if (note) {
      events.push([ts, `${p.srcIp}:${p.srcPort} -> ${p.dstIp}:${p.dstPort} -- ${note}`]);
    }
  }

  for (const f of detectExfiltration(packetsParsed)) {
    if (f.dst !== "10.0.0.50") {
      events.push([f.startTs, `EXFILTRATION: ${f.src} sent ${f.packetCount} packets ` +
        `(${f.totalBytes} bytes) to ${f.dst}:${f.port} -- likely data exfiltration to an external host`]);
    }
  }

  events.sort((a, b) => a[0] - b[0]);
  return events;
}

function main() {
  const path = process.argv[2] || "incident_node.pcap";
  const events = buildTimeline(path);
  console.log(`Incident timeline reconstructed from ${path}:\n`);
  const t0 = events.length > 0 ? events[0][0] : 0;
  for (const [ts, description] of events) {
    console.log(`[T+${(ts - t0).toFixed(2).padStart(7)}s] ${description}`);
  }
}

main();

File: module-18-network-forensics/incident_timeline.py
"""
Module 18 -- reconstructs a human-readable incident timeline from a
.pcap file, annotating each phase of an attack: reconnaissance, exploit
attempt, and exfiltration. This is Module 14's detection reasoning
(port-scan and large-transfer patterns) applied retrospectively to a
saved capture, instead of live/statistical detection -- the actual job of
an incident responder handed a pcap after the fact and asked "what
happened here, in order?"
"""
from collections import defaultdict

from pcap_reader import parse_frame, read_pcap

KNOWN_BAD_PAYLOADS = {
    b"smiley:)": "vsftpd 2.3.4 backdoor trigger string (CVE-2011-2523, same CVE as Module 16's database)",
}


def annotate_packet(pkt, parsed) -> str:
    for pattern, description in KNOWN_BAD_PAYLOADS.items():
        if pattern in parsed["payload"]:
            return f"EXPLOIT ATTEMPT -- payload matches known-bad pattern: {description}"
    return None


def detect_recon(packets_parsed, window_seconds=1.0, min_ports=4):
    by_src_dst = defaultdict(list)
    for ts, p in packets_parsed:
        if p["tcp_flags"] == 0x02:  # SYN only
            by_src_dst[(p["src_ip"], p["dst_ip"])].append((ts, p["dst_port"]))

    findings = []
    for (src, dst), events in by_src_dst.items():
        events.sort()
        ports = {port for ts, port in events if ts - events[0][0] <= window_seconds}
        if len(ports) >= min_ports:
            findings.append({
                "type": "reconnaissance", "src": src, "dst": dst,
                "ports_touched": sorted(ports), "start_ts": events[0][0],
            })
    return findings


def detect_exfiltration(packets_parsed, min_packets=10):
    by_src_dst_port = defaultdict(list)
    for ts, p in packets_parsed:
        if p["payload"]:
            by_src_dst_port[(p["src_ip"], p["dst_ip"], p["dst_port"])].append((ts, len(p["payload"])))

    findings = []
    for (src, dst, port), events in by_src_dst_port.items():
        if len(events) >= min_packets:
            total_bytes = sum(size for ts, size in events)
            findings.append({
                "type": "exfiltration", "src": src, "dst": dst, "port": port,
                "packet_count": len(events), "total_bytes": total_bytes, "start_ts": events[0][0],
            })
    return findings


def build_timeline(path: str):
    meta, packets = read_pcap(path)
    packets_parsed = [(pkt["ts"], parse_frame(pkt["frame"])) for pkt in packets]

    events = []

    recon_findings = detect_recon(packets_parsed)
    for f in recon_findings:
        events.append((f["start_ts"], f"RECONNAISSANCE: {f['src']} scanned {len(f['ports_touched'])} "
                                       f"ports on {f['dst']} ({f['ports_touched']})"))

    for ts, p in packets_parsed:
        note = annotate_packet((ts, p), p)
        if note:
            events.append((ts, f"{p['src_ip']}:{p['src_port']} -> {p['dst_ip']}:{p['dst_port']} -- {note}"))

    exfil_findings = detect_exfiltration(packets_parsed)
    for f in exfil_findings:
        if f["dst"] not in {"10.0.0.50"}:  # don't flag legitimate inbound-heavy traffic to the victim itself
            events.append((f["start_ts"], f"EXFILTRATION: {f['src']} sent {f['packet_count']} packets "
                                           f"({f['total_bytes']} bytes) to {f['dst']}:{f['port']} -- "
                                           f"likely data exfiltration to an external host"))

    events.sort(key=lambda e: e[0])
    return events


def main():
    import sys
    path = sys.argv[1] if len(sys.argv) > 1 else "incident.pcap"

    events = build_timeline(path)
    print(f"Incident timeline reconstructed from {path}:\n")
    t0 = events[0][0] if events else 0
    for ts, description in events:
        print(f"[T+{ts - t0:7.2f}s] {description}")


if __name__ == "__main__":
    main()
